File: elixir/lib/2018/03_no_matter_how_you_slice_it/aoc201803.ex
defmodule AOC2018.Day03 do
  @moduledoc """
  Advent of Code 2018, day 3: No Matter How You Slice It
  """
  require AOC

  @doc """
  Parse input
  """
  def parse(puzzle_input) do
    puzzle_input |> String.split("\n") |> Enum.map(&parse_claim/1) |> lay_out_claims()
  end

  @doc """
  Parse one claim.

  ## Examples:

      iex> parse_claim("#1 @ 258,327: 19x22")
      {1, {258, 327}, {19, 22}}
  """
  def parse_claim(claim) do
    [id, left, top, width, height] =
      claim
      |> String.split(["#", " @ ", ",", ": ", "x"], trim: true)
      |> Enum.map(&String.to_integer/1)

    {id, {left, top}, {width, height}}
  end

  @doc """
  Lay out all claims.

  ## Examples:

      iex> lay_out_claims([{1, {2, 1}, {2, 2}}, {2, {1, 2}, {2, 1}}])
      %{{2, 1} => [1], {2, 2} => [2, 1], {3, 1} => [1], {3, 2} => [1], {1, 2} => [2]}
  """
  def lay_out_claims(claims), do: lay_out_claims(claims, %{})
  def lay_out_claims([], fabric), do: fabric

  def lay_out_claims([{id, {left, top}, {width, height}} | claims], fabric) do
    lay_out_claims(
      claims,
      Enum.reduce(left..(left + width - 1), fabric, fn x, acc ->
        Enum.reduce(top..(top + height - 1), acc, fn y, acc ->
          Map.update(acc, {x, y}, [id], &[id | &1])
        end)
      end)
    )
  end

  @doc """
  Solve part 1
  """
  def part1(input), do: input |> overlapping_squares() |> Enum.count()

  @doc """
  Solve part 2
  """
  def part2(input), do: input |> nonoverlapping_claim()

  @doc """
  Identify overlapping squares.

  ## Examples:

      iex> overlapping_squares(%{{1, 2} => [2], {2, 1} => [1], {2, 2} => [2, 1]})
      [[2, 1]]
  """
  def overlapping_squares(squares), do: for({_, [_, _ | _] = ids} <- squares, do: ids)

  @doc """
  Identify claim that is not overlapping with other claims.

  ## Examples:

      iex> nonoverlapping_claim(%{{1, 1} => [3], {1, 2} => [2], {2, 1} => [1], {2, 2} => [2, 1]})
      3
  """
  def nonoverlapping_claim(squares) do
    all_claims = squares |> Map.values() |> Enum.concat() |> MapSet.new()
    overlapping_claims = squares |> overlapping_squares() |> Enum.concat() |> MapSet.new()

    MapSet.difference(all_claims, overlapping_claims) |> MapSet.to_list() |> hd()
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/2018/03_no_matter_how_you_slice_it/aoc201803.py
"""AoC 3, 2018: No Matter How You Slice It"""

# Standard library imports
import collections
import itertools
import pathlib
import sys

import parse

CLAIM_PATTERN = parse.compile("#{id:d} @ {col:d},{row:d}: {width:d}x{height:d}")


def parse(puzzle_input):
    """Parse input"""
    return dict(parse_claim(claim) for claim in puzzle_input.split("\n"))


def parse_claim(claim_text):
    """Parse one claim

    ## Example:

    >>> parse_claim("#7 @ 19,77: 2x3")
    (7, [(19, 77), (19, 78), (19, 79), (20, 77), (20, 78), (20, 79)])
    """
    claim = CLAIM_PATTERN.parse(claim_text)
    return (
        claim["id"],
        [
            (claim["col"] + dc, claim["row"] + dr)
            for dc in range(claim["width"])
            for dr in range(claim["height"])
        ],
    )


def part1(data):
    """Solve part 1"""
    overlaps = collections.Counter(itertools.chain(*data.values()))
    return sum(count > 1 for count in overlaps.values())


def part2(data):
    """Solve part 2"""
    overlaps = collections.Counter(itertools.chain(*data.values()))
    no_overlaps = {rect for rect, count in overlaps.items() if count == 1}
    return next(
        claim_id
        for claim_id, rects in data.items()
        if all(rect in no_overlaps for rect in rects)
    )


def solve(puzzle_input):
    """Solve the puzzle for the given input"""
    data = parse(puzzle_input)
    solution1 = part1(data)
    solution2 = part2(data)

    return solution1, solution2


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))
